Add missing tables to globally routed list in schema tracker only if they are not already present in a VSchema

rohit-nayak-ps opened the pull request:
## Description

Follows up on the idea in https://github.com/vitessio/vitess/pull/16517 by not considering tables already present in the globally routable list of tables.

See [RFC](https://github.com/vitessio/vitess/issues/17425) for design and implementation details.

## Related Issue(s)

https://github.com/vitessio/vitess/issues/16516

## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [ ] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [ ] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on CI?
-   [ ] Documentation was added or is not required

## Deployment Notes

<!-- Notes regarding deployment of the contained body of work. These should note any db migrations, etc. -->

GuptaManan100 commented on go/vt/vtgate/vindexes/vschema.go:
```diff
@@ -461,23 +461,27 @@ func (vschema *VSchema) AddUDF(ksname, udfName string) error {
 		if gt, ok := vschema.globalTables[tname]; ok {
 			// There is already an entry table stored in global tables
 			// with this name.
+			if !skipIfAlreadyGlobal {
+				// Called when updating from schema tracking
+				continue
+			}
```
The logic is fine, but doesn't the name of the variable seem exactly opposite of what it is doing? Like we are skipping adding to the global table if is already present when `skipIfAlreadyGlobal` is false? Am I misunderstanding this?

rohit-nayak-ps replied:
You are right! I changed the name of the variable from `override` to this, which changed its meaning. So I should reverse the values passed in. Will do that.

rohit-nayak-ps replied:
Fixed

harshit-gangal replied:
```suggestion
			if skipAlreadyGlobal {
				// Called when updating from schema tracking
				continue
			}
```

harshit-gangal replied:
`!` has to go I believe

rohit-nayak-ps replied:
Duh, yes, thanks, needed to change this as well since I switched the logic around. Fixed and tests are now passing.

GuptaManan100 reviewed (approved):
Just a nomenclature nit, otherwise looks good to me!

harshit-gangal reviewed (commented):
The issue with this approach arises when there are multiple unsharded keyspaces. This makes adding the table to global routing non-deterministic, failing to fully achieve the intended purpose of this PR.

rohit-nayak-ps commented:
> The issue with this approach arises when there are multiple unsharded keyspaces. This makes adding the table to global routing non-deterministic, failing to fully achieve the intended purpose of this PR.

You mean if there are tables of the same name in multiple unsharded keyspaces (and which are not in the VSchemas). 

We could collect all tables names which will be added like this (not in VSchema, but found by schema tracking) first and validate that there are no duplicates, marking ones with duplicates as ambiguous, while others get globally routed.

rohit-nayak-ps commented:
Discussed with @harshit-gangal . His point is that:

1. We start with a sharded keyspace `Sharded1` and one unsharded keyspace `Unsharded1`. `Unsharded1` has no vschema defined yet.
2. After this PR, schema tracking finds a table `ut1` in `Unsharded1` that is not in the current VSchema. It then adds it to the global tables list. User starts depending on `select * from ut1`.
3. Later user adds another unsharded keyspace `Unsharded2` which also has table named `ut1`. 
4. Now we will  mark `ut1` as `ambiguous` (this logic is not yet part of this PR, but is the enhancement I referred to in https://github.com/vitessio/vitess/pull/17371#issuecomment-2541328015)
5. Suddenly the user's  `select * from ut1` query will start erroring out. User won't have visibility as to why this has happened since they have not done anything obvious in Vitess like modifying/adding a VSchema and will have downtime until they drop the other table, which could be difficult to diagnose/fix.


Options here are:
1. Close this PR and not do anything
2. Merge this PR (with enhanced logic to detect ambiguous tables outside of current VSchemas), which will solve the problem for `X`% of customers who currently have downtime because we don't do this

I guess the question is what is `X`. Is it high enough to merge this. Or even if `X` is high, do we want to do this from principled viewpoint.

rohit-nayak-ps commented:
After discussions we decided to go ahead with the implementation as detailed in the [RFC](https://github.com/vitessio/vitess/issues/17425)

rohit-nayak-ps commented:
I have added a vtgate flag, `--mark_unique_unsharded_tables_as_global (default true)` to allow us the ability to disable the logic added in the PR in case some breaking changes have been introduced. I can remove it if reviewers feel it is not needed.

GuptaManan100 commented on go/vt/vtgate/vindexes/vschema_routing_test.go:
```diff
@@ -0,0 +1,481 @@
+		Tables: map[string]*vschemapb.Table{
+			"table5":   {}, // unique, should be added to global routing
+			"scommon1": {}, // common with unsharded1 and unsharded2, should be added to global routing because it's in the vschema
+			"scommon2": {}, // common with unsharded2, ambiguous because of sharded2
```
The comment seems wrong. It is not ambiguous because of sharded2 because sharded2 requires explicit routing. We even assert that `scommon2` can be globally routed.

GuptaManan100 commented on go/vt/vtgate/vindexes/vschema_routing_test.go:
```diff
@@ -0,0 +1,481 @@
+		// none should be considered for choice as global or ambiguous because of RequireExplicitRouting
+		Tables: map[string]*vschemapb.Table{
+			"table6":   {}, // unique
+			"scommon2": {}, // common with sharded2, but has RequireExplicitRouting
```
Should be common with sharded1.

GuptaManan100 commented on go/vt/vtgate/vindexes/vschema_routing_test.go:
```diff
@@ -0,0 +1,481 @@
+		Tables: map[string]*vschemapb.Table{
+			"table6":   {}, // unique
+			"scommon2": {}, // common with sharded2, but has RequireExplicitRouting
+			"scommon3": {}, // common with sharded2
```
Incorrect comment. `scommon3` is not common with anything. It is unique.

GuptaManan100 commented on go/test/endtoend/vreplication/r:
```diff
@@ -0,0 +1,50 @@
+cleanup() {
+    rm -rf /Users/rohit/vtdataroot/*
```
This looks wrong. Do we want to use an env variable `$VTDATAROOT` instead.

rohit-nayak-ps replied:
Mistakenly added my local test scripts. Removed.

GuptaManan100 commented on go/test/endtoend/vreplication/r2:
```diff
@@ -0,0 +1,17 @@
+cleanup() {
+    rm -rf /Users/rohit/vtdataroot/*
```
Same here

GuptaManan100 commented on go/test/endtoend/vreplication/global_routing_test.go:
```diff
@@ -0,0 +1,314 @@
+func TestGlobalRouting(t *testing.T) {
+	h := grHelpers{t}
+	exp := *h.getExpectations()
+	testCases := []grTestCase{
+		{unshardedHasVSchema: false, markAsGlobal: true},
+		{unshardedHasVSchema: false, markAsGlobal: false},
+		{unshardedHasVSchema: true, markAsGlobal: true},
+		{unshardedHasVSchema: true, markAsGlobal: false},
+	}
+	for _, tc := range testCases {
```
This feels duplicated. The `exp` already has this information in its keys. 
We could just do this instead - 
```go
for tc, funcs := range exp {
		require.NotNil(t, funcs)
		testGlobalRouting(t, tc.markAsGlobal, tc.unshardedHasVSchema, funcs)
	}
```
This avoids the need for explicitly writing testcases.

GuptaManan100 reviewed (approved):
Rest LGTM!

deepthi reviewed (commented):
1. Why does `BuildGlobalTables` need to be exported? I couldn't find any references to it outside the package.
2. Do we need an empty vschema json file? Can it be removed?
3. I assume you will be deleting the flag

harshit-gangal commented on go/vt/vtgate/vindexes/vschema_routing_test.go:
copyright header missing

harshit-gangal commented on go/vt/vtgate/vindexes/vschema_routing_test.go:
```diff
@@ -0,0 +1,482 @@
+	"vitess.io/vitess/go/vt/sqlparser"
+)
+
+func TestAutoGlobalRoutingExt(t *testing.T) {
```
nit: add a test comment

rohit-nayak-ps commented:
> 1. Why does `BuildGlobalTables` need to be exported? I couldn't find any references to it outside the package.
Unexported. It was used in a previous version of the PR.

> 2. Do we need an empty vschema json file? Can it be removed?
Removed

> 3. I assume you will be deleting the flag
Deleted

harshit-gangal commented on go/test/endtoend/vreplication/global_routing_test.go:
a test write up would help understand what cases are tested here and how is the setup

harshit-gangal commented on go/test/endtoend/vtgate/gen4/main_test.go:
```diff
@@ -100,6 +104,15 @@ func TestMain(m *testing.M) {
 			return 1
 		}
 
+		uKs2 := &cluster.Keyspace{
+			Name:      unsharded2Ks,
+			SchemaSQL: unsharded2SchemaSQL,
+		}
+		err = clusterInstance.StartUnshardedKeyspace(*uKs2, 0, false)
+		if err != nil {
+			return 1
+		}
+
```
we should add comment here to understand why this keyspace was added. Otherwise, without adding a new test it remains unclear.

rohit-nayak-ps commented on go/vt/vtgate/vindexes/vschema_routing_test.go:
```diff
@@ -0,0 +1,500 @@
+	mustNotRouteGlobally(t, "table6")
+	mustNotRouteGlobally(t, "scommon3")
+}
+
```
The next test is moved over from `vschema_test.go` so that it is colocated with the new global routing tests.

rohit-nayak-ps commented:
> I have added a vtgate flag, `--mark_unique_unsharded_tables_as_global (default true)` to allow us the ability to disable the logic added in the PR in case some breaking changes have been introduced. I can remove it if reviewers feel it is not needed.

This has been now removed.
